### ISSUE TITLE
Contour polygon insideness is reverse.

### ISSUE THREAD

**Issue by dongli** [d3/d3-contour#12]
Thanks for creating this great library! It saves me a lot of time!

I have applied it to create contours for precipitation data which is regional not global, and projected the contours onto orthogonal projection, but I found that the polygon insideness of some levels is reverse. In following figure, the left panel is wrong. Any idea why `d3` draws this path with its insideness reverse? When not projected, there is no such problem.

<img width="963" alt="default" src="https://user-images.githubusercontent.com/367959/27386332-1b7fe4c2-565b-11e7-914c-64c36313900f.png">

```javascript
...
var contours = d3Contour.contours().size([record.width, record.height])
                                   .smooth(false)
                                   .thresholds(record.colormap.domain)(record.rawData)
                                   // Change planar coordinates to spherical coordinates.
                                   .map(convertCoordinates)

svg.selectAll('path')
   .data(contours)
   .enter().append('path')
   .exit().remove()
svg.selectAll('path')
   .data(contours)
   .attr('d', path)
   .attr('transform', view.transform)
   .attr('fill', (d, i) => record.colormap.range[i])
   .attr('fill-rule', 'evenodd')
        .attr('value', d => d.value)

function convertCoordinates (d) {
  var e = Object.assign({}, d)
  e.coordinates = d.coordinates.map(polygon => {
    return polygon.map(ring => {
      return ring.map(point => {
        return [
          point[0] * record.grid.dlon + record.grid.startLon,
          point[1] * record.grid.dlat + record.grid.startLat
        ]
      })
    })
  })
  return e
}
```

**Comment by mbostock** [comment]
I think you probably want to reverse the coordinates of the rings, since the GeoJSON polygons generated by d3-contour are planar, while the polygons you want to pass to the orthographic projection are assumed to be spherical; the assumed coordinate system for planar geometry assumes that *y* points down, while the assumed coordinate system for spherical geometry assumes that *y* (*latitude*) points up.

(It’s also *possible* that there’s a winding order bug in d3-geo for very small polygons, but that should be unlikely given that the size of the polygons here is determined by the size of the pixel grid, and the marching squares algorithm is guaranteed not to generated degenerate polygons.)

But, you haven’t linked to a live example that reproduces the issue, so it’s impossible for me to investigate. When reporting an issue, please include a link to a live example, preferably on [bl.ocks.org](http://bl.ocks.org) or [RunKit](https://runkit.com/npm/d3), or as a pull request in the [appropriate repository](https://github.com/d3) with new unit tests, that demonstrates that the described behavior is not the [expected behavior](https://github.com/d3/d3/blob/master/API.md). Use only the minimum amount of code necessary to reproduce the unexpected behavior.

A good bug report should **isolate specific methods** that exhibit unexpected behavior and precisely define how expectations were violated. What did you expect the method or methods to do, and how did the observed behavior differ? The more precisely you isolate the issue, the faster I can investigate.

Non-actionable bugs may be temporarily closed until you provide additional information. If you can’t isolate the bug any further, please let me know and I will try to help. However, I beg your patience: I must balance your request against many other responsibilities—fixing other bugs, answering other questions, new features, new documentation, *etc.* Please also consider asking for help on [Stack Overflow](https://stackoverflow.com/tags/d3.js).

If you have a question about D3’s behavior and want to discuss it with other users, also consider the [d3-js Google Group](https://groups.google.com/forum/#!forum/d3-js) or [joining](https://d3-slackin.herokuapp.com/) the [d3-js Slack](https://d3js.slack.com/).

Thank you! 🤗

**Comment by dongli** [comment]
Thanks for help! I have encountered difficulties to extract an actionable code snippet, since it involves several data and other codes. It would be great if you can guide me to figure out where is wrong.

I have converted the planar coordinates into spherical ones, so it may not be the problem of coordinates? And when used `d3.geoPath(d3.geoIdentity())` path, the wrong level above is correct but with upside down and scale not matched. See the following figure, where for the convenience, I have flipped the right panel in Y direction.
<img width="945" alt="default" src="https://user-images.githubusercontent.com/367959/27392211-b3a7115c-566b-11e7-90c2-e3ee883ce3fc.png">

So does it mean it is the problems of projection (`geoOrthogonal`)?

**Comment by mbostock** [comment]
Sorry but I cannot investigate further unless you post code that reproduces the issue. I also suggest taking a look at the d3.geoStitch example linked from the README that demonstrates more generally steps that may be required to convert planar to spherical coordinates; it is not simply a matter of scaling and translating.

**Comment by dongli** [comment]
@mbostock I may have found the cause. In the first figure (thank [geojson.iso](http://geojson.io/)), I spotted a polygon with negative area, which means it should be a hole, but it is outside the main polygon (the biggest one). After delete all the hole polygons, the projected contour has correct fill as shown in second figure. Any idea why this lost polygon is generated?

<img width="478" alt="default" src="https://user-images.githubusercontent.com/367959/27414379-afc0f76e-56c6-11e7-957d-0dbc4da355bb.png">

![2017-06-22 10 16 30](https://user-images.githubusercontent.com/367959/27414418-ef6b3438-56c6-11e7-84d7-e8a4bb53e369.png)

**Comment by mbostock** [comment]
Sorry but I cannot investigate further unless you post code that reproduces the issue.

**Comment by dongli** [comment]
I have created an example to show the problem, but I cannot upload binary data to `gist`... so I created tar file [here](https://www.dropbox.com/s/ipna6j3xtlu3m5c/debug-contour.tgz?dl=0).

**Comment by dongli** [comment]
I inspected more into the data, and found the data that causes the negative area polygon is as following:

![2017-06-22 12 15 27](https://user-images.githubusercontent.com/367959/27417135-b288736c-56d7-11e7-9d85-0dd5a27cfbd8.png)

**Comment by dongli** [comment]
@mbostock I think I found a bug in `d3-contour`:

```javascript
holes.forEach(function(hole) {
  for (var i = 0, n = polygons.length, polygon; i < n; ++i) {
    if (contains((polygon = polygons[i])[0], hole) !== -1) { // This should be compared with -1!
      polygon.push(hole);
      return;
    }
  }
});
```
`contains` function return `-1` to indicate not-contain situation, but that `if` condition pass `-1`! When add `!== -1`, the contours are correct.

**Comment by mbostock** [comment]
Thanks for the test case! I’ve verified your fix in a test and released it as 1.1.1. Here’s a before using your data:

![untitled 22](https://user-images.githubusercontent.com/230541/27509567-ec3c7124-58b4-11e7-92e8-5b4901c35f81.png)

And after:

![untitled 23](https://user-images.githubusercontent.com/230541/27509568-f04e1aec-58b4-11e7-839a-980b7c527ca7.png)

**Comment by dongli** [comment]
@mbostock I am very happy to contribute my humble efforts. :stuck_out_tongue:
